node: add channel read/write type constraints

tbjump opened the pull request:
there are no functional changes in this PR. 

I am adding read/write type constraints to the main channels defined in cmd/guardiand/node.go. This makes the code easier to audit.

jynnantonix commented on node/cmd/guardiand/node.go:
```diff
@@ -771,36 +772,89 @@ func runNode(cmd *cobra.Command, args []string) {
+	// Per-chain msgC
+	chainMsgC := make(map[vaa.ChainID]chan *common.MessagePublication)
+	// aggregate per-chain msgC into msgC.
+	// SECURITY defense-in-depth: This way we enforce that a watcher must set the msg.EmiggerChain to its chainId.
```
"msg.EmiggerChain" -> "msg.EmitterChain"

jynnantonix commented on node/cmd/guardiand/node.go:
```diff
@@ -771,36 +772,89 @@ func runNode(cmd *cobra.Command, args []string) {
+					msgWriteC <- msg
+				} else {
+					// SECURITY: This should never happen. If it does, a watcher has been compromised.
+					logger.Error("SECURITY CRITICAL: Received observation from a chain that was not marked as originating from that chain",
```
Should this be `logger.Fatal` rather than just an error?

tbjump replied:
thought about this as well. On the one hand, this is a failure that definitely needs to be looked at. On the other hand, other parts of the software could continue running fine. I think in the interest of catching bugs early during development, I can change this to logger.Fatal. What do you think?

jynnantonix commented on node/cmd/guardiand/node.go:
```diff
@@ -771,36 +772,89 @@ func runNode(cmd *cobra.Command, args []string) {
+	for chainId := range sdk.KnownTokenbridgeEmitters { // TODO maybe not use KnownTokenbridgeEmitters here and think about adding just a list of chainIds to the sdk mainnet_consts.go
+		chainMsgC[chainId] = make(chan *common.MessagePublication)
+		go func(c <-chan *common.MessagePublication, chainId vaa.ChainID) {
+			for msg := range c {
```
This will leak the goroutine, which isn't that big of an issue but if we want to do it the Right Way (TM) then we should select on the root context here:

```go
for {
	select {
		case <-rootCtx.Done():
			return
		case msg := <-c:
			// Handle msg
	}
}
```

jynnantonix reviewed (approved):
Overall idea looks good to me.

tbjump commented on node/cmd/guardiand/node.go:
```diff
@@ -771,36 +772,89 @@ func runNode(cmd *cobra.Command, args []string) {
 
-	// Injected VAAs (manually generated rather than created via observation)
-	injectC := make(chan *vaa.VAA)
+	// the following channels are created in a block such that they can only be accessed through their explicit read/write aliases
```
did we update the go version and can use generics now?

tbjump replied:
Looks like guardiand is still using go1.17? https://github.com/wormhole-foundation/wormhole/blob/f60e6669ec5e65ca2390a6ac10b834d2dec5821a/.github/workflows/build.yml#L254. Would love to see it upgraded though.

hendrikhofstadt replied:
Josh seems to be tackling it #1934

kcsongor replied:
I really like the `channelPair` idea. On the type inference point, even if go will infer the types of those channels, I think we should still explicitly supply the type instantiation here. In my experience, relying on inference for return type polymorphic functions leads to subtle refactoring hazards and overall loss of readability.

tbjump commented:
refactored the channel pair creation using go generics. This requires go > 1.18 which is tackled in https://github.com/wormhole-foundation/wormhole/issues/1934. Will rebase once that's done but until then the tests for this PR will probably fail.

hendrikhofstadt commented:
@tbjump pls rebase

tbjump commented:
rebased and resolved conflicts. Also added `vaa.GetAllNetworkIDs()` which may be useful in other parts of the codebase. If there is an existing way to get all network IDs, that I did not find, please let me know!

tbjump commented:
Codeowner review required: 
* @bruce-riley or @jynnantonix 
* @bruce-riley or @claudijd 
* @leoluk or @hendrikhofstadt 
* @jynnantonix or @hendrikhofstadt 
* @evan-gray or @jumpsiegel 
* @jumpsiegel 
* @evan-gray or @hendrikhofstadt 
* @bruce-riley or @hendrikhofstadt 
* @hendrikhofstadt 

Smallest sufficient set of reviewers: @hendrikhofstadt, @jumpsiegel, @bruce-riley

jynnantonix commented on sdk/vaa/structs.go:
```diff
@@ -265,6 +265,38 @@ func ChainIDFromString(s string) (ChainID, error) {
+		ChainIDBtc,
+		ChainIDInjective,
+		ChainIDNeon,
+		ChainIDWormchain,
```
Can we keep this list in the same order as the consts below?  That will make it easier to see if any are missing.

tbjump replied:
done

hendrikhofstadt commented on node/cmd/guardiand/node.go:
```diff
@@ -811,36 +811,64 @@ func runNode(cmd *cobra.Command, args []string) {
+				return
+			case msg := <-c:
+				if msg.EmitterChain == chainId {
+					msgWriteC <- msg
```
this could do a context aware write

tbjump replied:
can you explain / show me how?

tbjump replied:
are you ok with the latest version?

bruce-riley commented on node/cmd/guardiand/node.go:
```diff
@@ -811,36 +811,64 @@ func runNode(cmd *cobra.Command, args []string) {
+	chainMsgC := make(map[vaa.ChainID]chan *common.MessagePublication)
+	// aggregate per-chain msgC into msgC.
+	// SECURITY defense-in-depth: This way we enforce that a watcher must set the msg.EmitterChain to its chainId, which makes the code easier to audit
+	for _, chainId := range vaa.GetAllNetworkIDs() {
```
I am not convinced that this check is worth the additional overhead and complexity, but I won't block the PR over it.

jynnantonix reviewed (dismissed):
I'm assuming this PR is going to get squashed into a single commit.

evan-gray commented:
@tbjump this has conflicts - converting to draft for now, but please address

tbjump commented:
resolved merge conflicts.

jynnantonix reviewed (approved):
lgtm % Hendrik's comment

tbjump commented:
@hendrikhofstadt @jumpsiegel @bruce-riley resolved all comments, resolved merge conflicts and passing all tests now. Please review again.

hendrikhofstadt reviewed (commented):
Still has an open comment and needs a rebase

tbjump commented:
resolved merge conflicts again

bruce-riley commented:
While resolving merge conflicts in node.go, can you please change this to use the new approach?

tbjump commented:
resolved merge conflicts again and addressed @bruce-riley 's todo. Please review once more.
